File: src/main/scala/g0701_0800/s0739_daily_temperatures/Solution.scala
package g0701_0800.s0739_daily_temperatures

// #Medium #Top_100_Liked_Questions #Array #Stack #Monotonic_Stack #Programming_Skills_II_Day_6
// #Big_O_Time_O(n)_Space_O(n) #2023_11_09_Time_931_ms_(97.30%)_Space_69.2_MB_(97.30%)

object Solution {
    def dailyTemperatures(temperatures: Array[Int]): Array[Int] = {
        val sol = new Array[Int](temperatures.length)
        sol(temperatures.length - 1) = 0
        for (i <- (sol.length - 2) to 0 by -1) {
            var j = i + 1
            while (j < sol.length) {
                if (temperatures(i) < temperatures(j)) {
                    sol(i) = j - i
                    j = sol.length // exit loop
                } else {
                    if (sol(j) == 0) {
                        j = sol.length // exit loop
                    } else {
                        j += sol(j)
                    }
                }
            }
        }
        sol
    }
}

File: src/main/rust/g0701_0800/s0739_daily_temperatures/Solution.rs
// #Medium #Top_100_Liked_Questions #Array #Stack #Monotonic_Stack #Programming_Skills_II_Day_6
// #Big_O_Time_O(n)_Space_O(n) #2024_09_13_Time_33_ms_(91.80%)_Space_3.4_MB_(99.45%)

impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut sol = vec![0; temperatures.len()];
        for i in (0..temperatures.len() - 1).rev() {
            let mut j = i + 1;
            while j < temperatures.len() {
                if temperatures[i] < temperatures[j] {
                    sol[i] = (j - i) as i32;
                    break;
                } else if sol[j] == 0 {
                    break;
                }
                j += sol[j] as usize;
            }
        }
        sol
    }
}
